include/up-cpp/uri/builder/BuildUResource.h

```diff
@@ -0,0 +1,139 @@
+#include <string_view>
+#include <arpa/inet.h>
+#include <spdlog/spdlog.h>
+#include <up-cpp/uri/tools/Utils.h>
+#include "up-core-api/uri.pb.h"
+
+namespace uprotocol::uri {
+    class BuildUResource {
+        uprotocol::v1::UResource resource_;
+
+    public:
+        BuildUResource();
+
+        /**
+         * set name of the resource. if name is empty or blank, an error will
+         * be logged and the name will not be set
+         *
+         * if name is already set, it will not be changed
+         *
+         * @param name
+         * @return
+         */
+        auto setName(const std::string &name) -> BuildUResource &;
+
+        /**
+         * set instance of the resource. if instance is empty or blank, an
+         * error will be logged and the instance will not be set
+         *
+         * if instance is already set, it will not be changed   
+         *
+         * @param instance 
+         * @return 
+         */
+        auto setInstance(const std::string &instance) -> BuildUResource &;
+
+        /**
+         * set message of the resource. if message is empty or blank, an error
+         * will be logged and the message will not be set
+         *
+         * if message is already set, it will not be changed
+         *
+         * @param message
+         * @return
+         */
+        auto setMessage(const std::string &message) -> BuildUResource &;
+
+        /**
+         * set id of the resource. if id is 0, an error will be logged and the
+         * id will not be set
+         *
+         * if id is already set, it will not be changed
+         *
+         * @param id
+         * @return
+         */
+        auto setID(const uint32_t &id) -> BuildUResource &;
+
+        /**
+         * set id of the resource. if id is 0, an error will be logged and the
+         * id will not be set
+         *
+         * if id is already set, it will not be changed
+         *
+         * @param id
+         * @return
+         */
+        auto setRpcRequest(const std::string &method) -> BuildUResource &;
+
+        /**
+         * set Rpc request with both method and id. a wrapper method on
+         * setName, setId and set instance
+         *
+         * @param method 
+         * @param id 
+         * @return 
+         */
+        auto setRpcRequest(
+                const std::string &method, uint32_t id) -> BuildUResource &;
+
+        /**
+         * set Rpc request with id. a wrapper method on setName, setId
+         *
+         * @param method 
+         * @param id 
+         * @return 
+         */
+        auto setRpcRequest(const uint32_t id) -> BuildUResource &;
+
+        /**
+         * set RPC response 
+         * @return 
+         */
+        auto setRpcResponse() -> BuildUResource &;
+
+        /**
+         * 
+         * @return uprotocol::v1::UResource
+         */
+        auto build() const -> uprotocol::v1::UResource;
+    };
+    
+    [[nodiscard]] [[maybe_unused]] auto static isRPCMethod(
```

Comment:
setID() should be setId()?

Reply:
> Builders should not validate, validators are separate

I think the way that `UMessageBuilder` is written to produce a _valid_ UMessage, it makes sense that it does validation and if fails validation an Error is returned in the Result, e.g. see [here](https://github.com/eclipse-uprotocol/up-rust/blob/1bb08ba7a3666e58c316489fbcf3da3e29dee611/src/umessage/umessagebuilder.rs#L617).

But that's because the API is validating at the UMessage creation point. I don't like the idea of validating on the subcomponents of a UMessage.